### PR TITLE
ci: move dependabot updating process to Gtihub Actions

### PR DESCRIPTION
**Pull request by yihau**
#### Problem

our `ci/dependabot-updater.sh` broke recently because dependabot don't add "bump aaa from bbb to ccc" to git log anymore

#### Summary of Changes

refactor and move it to Github Actions

### CODE REVIEW

File: .github/workflows/dependabot-pr.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+          new_version="${BASH_REMATCH[3]}"
+          echo "crate_name: $crate_name, old_version: $old_version, new_version: $new_version"
+
+          ./scripts/cargo-for-all-lock-files.sh -- update -p $crate_name:$old_version --precise $new_version
+          ./scripts/cargo-for-all-lock-files.sh -- tree
+
+      - name: push
+        run: |
+          git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
+          git config user.name "github-actions[bot]"
```
Comment (ryoqun) [inline .github/workflows/dependabot-pr.yml]: custom name and email aren't allowed? say, "dependabot-github-action". while `41898282` is still `git grep`-able from git log to reach to this github workflow file, i think more descriptive name helps.

File: .github/workflows/dependabot-pr.yml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+          git add .
```
Comment (ryoqun) [inline .github/workflows/dependabot-pr.yml]: I'd prefer globing like `:**/Cargo.lock` as before.

anyone can edit `.gitignore`, a bit of extreme thought. but, if we're ever tricked into writing credentials into some random file, it's over.

**Comment by ryoqun** [comment]
btw, thanks for taking over my baby... solana-labs/solana#9180

**Comment by ryoqun** [comment]
also, i wonder we could use this new feature to kill this thing entirely: https://github.com/dependabot/dependabot-core/issues/1190#issuecomment-1693360104 (haha nice old me; i left tracking upstream issue in the solana-labs/solana#9180.

File: .github/workflows/dependabot-pr.yml (hunk 1 above)
Comment (yihau) [reply .github/workflows/dependabot-pr.yml]: oh okay! maybe I switch them back to the original ones! 
https://github.com/anza-xyz/agave/blob/9035690e40c0dbffc83ef8c2aa168078853359d7/ci/dependabot-pr.sh#L23
https://github.com/anza-xyz/agave/blob/9035690e40c0dbffc83ef8c2aa168078853359d7/ci/dependabot-pr.sh#L29

File: .github/workflows/dependabot-pr.yml (hunk 2 above)
Comment (yihau) [reply .github/workflows/dependabot-pr.yml]: yeah... you're right. let me test it a bit

**Comment by yihau** [comment]
let me check the new feature!

**Comment by yihau** [comment]
if I understand correctly, the topic is on the "gruop" feature. it looks like it will combine many dependencies as possible in a single PR 🤔 which I think is not good for us. I guess splitting them into different ones is more easy for us to bisect regression or something bad.

File: .github/workflows/dependabot-pr.yml (hunk 1 above)
Comment (yihau) [reply .github/workflows/dependabot-pr.yml]: what do you think if we just use the dependabot itself :trollface: [14a411d](https://github.com/anza-xyz/agave/pull/1466/commits/14a411d40c05e3be3a352ac76f51fffc622fa1b9)

File: .github/workflows/dependabot-pr.yml (hunk 2 above)
Comment (yihau) [reply .github/workflows/dependabot-pr.yml]: fixed: https://github.com/anza-xyz/agave/pull/1466/commits/0207a164f03255ee445647b4c9bc1af98937279a

**Comment by CriesofCarrots** [comment]
I'll defer to Ryo here.

File: .github/workflows/dependabot-pr.yml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            exit 1
+          fi
+
+          crate_name="${BASH_REMATCH[1]}"
+          old_version="${BASH_REMATCH[2]}"
+          new_version="${BASH_REMATCH[3]}"
+          echo "crate_name: $crate_name, old_version: $old_version, new_version: $new_version"
+
+          ./scripts/cargo-for-all-lock-files.sh -- update -p $crate_name:$old_version --precise $new_version
+          ./scripts/cargo-for-all-lock-files.sh -- tree
```
Comment (ryoqun) [inline .github/workflows/dependabot-pr.yml:36]: nit: i don't think `tree` is needed here. the preceding `update` should have updated any `Cargo.lock`s...

File: .github/workflows/dependabot-pr.yml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+      - name: push
+        run: |
+          if [ -z "$(git status --porcelain)" ]; then
+            echo "✅ Nothing to update"
+          else
+            git config user.email "49699333+dependabot[bot]@users.noreply.github.com"
+            git config user.name "dependabot[bot]"
+            git add **/Cargo.lock
+            git commit -am "update all Cargo files"
```
Comment (ryoqun) [inline .github/workflows/dependabot-pr.yml:46]: nit: I prefer title-case here by using upper-case `U`: `"Update all Cargo files"`.

**Review by ryoqun — APPROVED** [review]
lgtm with nits

**Comment by yihau** [comment]
thank you! let me merge this one and test it first. I will have another PR for those two nits!

File: .github/workflows/dependabot-pr.yml (hunk 4 above)
Comment (yihau) [reply .github/workflows/dependabot-pr.yml:46]: fixed: [4e5cb5a](https://github.com/anza-xyz/agave/pull/1496/commits/4e5cb5a5a5075db342538cab798893916e637a23)

File: .github/workflows/dependabot-pr.yml (hunk 3 above)
Comment (yihau) [reply .github/workflows/dependabot-pr.yml:36]: actually we will need to do both strict rule and relaxed rule. the first is for multiple version of crates in our repo. the second is for crates that are not in the `programs/sbf/Cargo.toml`. 

https://github.com/anza-xyz/agave/pull/1496/commits/2898cc094c030bf60dcc73b2721e735e5d2ddec2

File: ci/dependabot-updater.sh [hunk 5]
```diff
@@ -1,35 +0,0 @@
...
-  for lock in $(git grep --files-with-matches '^name = "'$package'"$' :**/Cargo.lock); do
-    # it's possible our current versions are out of sync across lock files,
-    # in that case try to sync them up with $relaxed_parsed_update_args
-    _ scripts/cargo-for-all-lock-files.sh \
-      "$lock" -- \
-      update $parsed_update_args ||
-      _ scripts/cargo-for-all-lock-files.sh \
-        "$lock" -- \
-        update $relaxed_parsed_update_args
-  done
```
Comment (ryoqun) [inline ci/dependabot-updater.sh:32]: here (backref: https://github.com/anza-xyz/agave/pull/1498/files#r1616475249)

**Pull request merged**